virtio-spec/src/virtq/alloc.rs

```diff
@@ -0,0 +1,96 @@
+use ::alloc::alloc::handle_alloc_error;
+use allocator_api2::alloc::{AllocError, Allocator, Global};
+use allocator_api2::boxed::Box;
+
+use super::*;
+
+impl Avail {
+    pub fn new(queue_size: u16, has_event_idx: bool) -> Box<Self> {
+        Self::new_in(queue_size, has_event_idx, Global)
+    }
+
+    pub fn try_new(queue_size: u16, has_event_idx: bool) -> Result<Box<Self>, AllocError> {
+        Self::try_new_in(queue_size, has_event_idx, Global)
+    }
+
+    pub fn new_in<A: Allocator>(queue_size: u16, has_event_idx: bool, alloc: A) -> Box<Self, A> {
+        Self::try_new_in(queue_size, has_event_idx, alloc)
+            .unwrap_or_else(|_| handle_alloc_error(Self::layout(queue_size, has_event_idx)))
+    }
+
+    pub fn try_new_in<A: Allocator>(
+        queue_size: u16,
+        has_event_idx: bool,
+        alloc: A,
+    ) -> Result<Box<Self, A>, AllocError> {
+        let layout = Self::layout(queue_size, has_event_idx);
+
+        let mem = alloc.allocate_zeroed(layout)?;
+        let mem = NonNull::slice_from_raw_parts(mem.cast(), layout.size());
+        let raw = Self::from_ptr(mem).unwrap();
+        let boxed = unsafe { Box::from_raw_in(raw.as_ptr(), alloc) };
+
+        debug_assert_eq!(Layout::for_value(&*boxed), layout);
+        debug_assert_eq!(boxed.ring(has_event_idx).len(), queue_size.into());
+        debug_assert_eq!(boxed.used_event(has_event_idx).is_some(), has_event_idx);
+
+        Ok(boxed)
+    }
+}
+
+impl Used {
+    pub fn new(queue_size: u16, has_event_idx: bool) -> Box<Self> {
+        Self::new_in(queue_size, has_event_idx, Global)
+    }
+
+    pub fn try_new(queue_size: u16, has_event_idx: bool) -> Result<Box<Self>, AllocError> {
+        Self::try_new_in(queue_size, has_event_idx, Global)
+    }
+
+    pub fn new_in<A: Allocator>(queue_size: u16, has_event_idx: bool, alloc: A) -> Box<Self, A> {
+        Self::try_new_in(queue_size, has_event_idx, alloc)
+            .unwrap_or_else(|_| handle_alloc_error(Self::layout(queue_size, has_event_idx)))
+    }
+
+    pub fn try_new_in<A: Allocator>(
+        queue_size: u16,
+        has_event_idx: bool,
+        alloc: A,
+    ) -> Result<Box<Self, A>, AllocError> {
+        let layout = Self::layout(queue_size, has_event_idx);
+
+        let mem = alloc.allocate_zeroed(layout)?;
+        let mem = NonNull::slice_from_raw_parts(mem.cast(), layout.size());
```

Comment:
Is `Used` padded at the end? If I understand correctly, this line will provide a slice with the padding included and I think the code in mod.rs is written with the assumption that the slice does not include any padding.

Reply:
This is padded, since `layout` is padded. I think `Used::from_ptr` still works fine?

Reply:
You are right